src/collapse/perf.rs

```diff
@@ -143,27 +137,52 @@ impl Default for Folder {
     }
 }
 
-impl Collapse for Folder {
-    fn collapse<R, W>(&mut self, reader: R, writer: W) -> io::Result<()>
+impl CollapsePrivate for Folder {
+    fn pre_process<R>(&mut self, reader: &mut R, occurrences: &mut Occurrences) -> io::Result<()>
     where
-        R: BufRead,
-        W: Write,
+        R: io::BufRead,
     {
-        // Do collapsing...
-        if self.occurrences.is_concurrent() {
-            self.collapse_multi_threaded(reader)?;
-        } else {
-            self.collapse_single_threaded(reader)?;
+        // If user has provided an event filter, do nothing...
+        if self.event_filter.is_some() {
+            return Ok(());
         }
 
-        // Write results...
-        self.occurrences.write_and_clear(writer)?;
+        // Otherwise, we don't know what the event filter should be; so process
+        // the first stack to figure it out (the worker threads need this
+        // information to get started). Only read one stack, however, as we would
+        // like the remaining stacks to be processed on the worker threads.
+        let mut line_buffer = String::new();
+        self.process_stack(&mut line_buffer, reader, occurrences)?;
```

Comment:
Ah, interesting, so this will drop a single sample from the input if I'm reading it right?

Reply:
Correct.